bigtable/client/data_client.cc

```diff
@@ -0,0 +1,70 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/data_client.h"
+
+namespace btproto = ::google::bigtable::v2;
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Implement a simple DataClient.
+ *
+ * This implementation does not support multiple threads, handle GOAWAY
+ * responses, nor does it refresh authorization tokens.  In other words, it is
+ * extremely bare bones.
+ */
+class DefaultDataClient : public DataClient {
+ public:
+  DefaultDataClient(std::string project, std::string instance,
+                    ClientOptions options)
+      : project_(std::move(project)),
+        instance_(std::move(instance)),
+        credentials_(options.credentials()),
+        channel_(grpc::CreateChannel(options.data_endpoint(),
+                                     options.credentials())),
+        bt_stub_(google::bigtable::v2::Bigtable::NewStub(channel_)) {}
+
+  DefaultDataClient(std::string project, std::string instance)
+      : DefaultDataClient(std::move(project), std::move(instance),
+                          ClientOptions()) {}
+
+  std::string const& ProjectId() const override;
```

Comment:
This should be `project_id()`, no? It's a simple accessor. I see it's inherited; any reason why it's like this in the parent class? Is it because it *could* be complex rather than a simple accessor? Seems like it shouldn't be.

Reply:
Fixed.

It should not be in the parent class because that would make the parent not an interface (or pure abstract).  And we want the class to be pure abstract so we can mock it completely.